engine/platform/posix/sys_posix.c

```diff
@@ -169,3 +182,63 @@ void Platform_Sleep( int msec )
 	usleep( msec * 1000 );
 }
 #endif // XASH_TIMER == TIMER_POSIX
+
+static void *Sys_TimerThread( void *v )
+{
+	struct timespec time, time1, time2, time3 = { 0,0 };
+	time.tv_sec = 0;
+	time.tv_nsec = g_timer.frametime * 1e9;
+	
+	while (1)
+	{
+		long long tdiff;
+		nanosleep(&time, &time1);
+		//printf("%ld\n", time1.tv_nsec );
+		time1.tv_sec = 0;
+		time1.tv_nsec = 0;
+		clock_gettime(CLOCK_MONOTONIC, &time2);
+		tdiff = time2.tv_nsec - time3.tv_nsec - g_timer.frametime * 1e9;
+		if (tdiff < 0)
+			tdiff = 0;
+		if (tdiff > g_timer.frametime / 2.0 * 1e9)
+			tdiff = 0;
+
+		pthread_mutex_lock(&g_timer.tmutex);
+
+		time.tv_nsec = 1e9 * g_timer.frametime - tdiff;
+		if (g_timer.flag == 0)
+			pthread_cond_signal(&g_timer.tsignal);
+		g_timer.flag = 1;
+		pthread_mutex_unlock(&g_timer.tmutex);
+		time3 = time2;
+	}
+	return 0;
+}
+
+void Platform_Delay( double time )
+{
+	if (g_timer.flag)
+	{
+		g_timer.flag = 0;
+		return;
+	}
+
+	pthread_mutex_lock(&g_timer.tmutex);
+	g_timer.frametime = time;
+	while (!g_timer.flag)
+		pthread_cond_wait(&g_timer.tsignal, &g_timer.tmutex);
+	g_timer.flag = 0;
+	pthread_mutex_unlock(&g_timer.tmutex);
+}
+
+void Platform_TimerInit( void )
+{
+	g_timer.flag = 1;
+	g_timer.frametime = 1.0 / host_maxfps->value;
```

Comment:
What if maxfps value is zero? Will it just run with +infinity?

Otherwise, what if this is a dedicated server? Dedicated servers don't depend on fps_max, but on sys_ticrate instead. I'm not sure if dedicated servers need this timer code at all, they're servers after all.

Reply:
I think it's better to replace this with some hardcoded value, it's just for first frame, anyway